### ISSUE TITLE
Watch-dog для GC

### ISSUE THREAD

**Issue by aurusov** [LeKaitoW/raox#112]
Запуск вот такой модели приводит к вечному циклу

``` antlr
event ClientInput() {
    ClientInput.plan(currentTime + 1);
}

set init() {
    ClientInput.plan(0);
}
```

Это нормально. Модель отлично отстанавливается по команде `Model/Stop`. Но если включить мониторинг, то среда зависает. На остановку не реагирует.

Предлагается добавить [watch-dog](https://ru.wikipedia.org/wiki/%D0%A1%D1%82%D0%BE%D1%80%D0%BE%D0%B6%D0%B5%D0%B2%D0%BE%D0%B9_%D1%82%D0%B0%D0%B9%D0%BC%D0%B5%D1%80) для `GC`. При достижении неких порогов (по памяти или кол-ву объектов или времени реакции или еще как) надо стопить модель **автоматически** с сообщением об ошибке.

Можно повторить через процессную модель https://github.com/aurusov/raox/issues/358

**Comment by bogachev-pa** [comment]
Конечно, зависает. Мы в бесконечном цикле добавляем записи о совершении события в базу данных и очень скоро у нас кончается память. Попробуйте запустить в режиме `normal speed` - все будет ок.

**Comment by aurusov** [comment]
> Конечно, зависает.

Это в корне неправильное понимание работы любой IDE. Запущенный из неё процесс может быть прибит в любое время. А у нас запущенный процесс прибивает IDE. Может приведешь аналоги, которые ведут себя так же ?

**Comment by bogachev-pa** [comment]
В таком случае неправильное понимание работы `IDE` не у меня, а у разработчиков `eclipse` - потому что я могу руками написать такой код на джаве и `eclipse` зависнет. Своей `IDE`-то у нас нет теперь.

Конечно, ничто не мешает нам не ровняться на эклипс и попытаться своими силами эту проблему для нашего случая решить. У вас есть идеи? Мне пока приходила в голову только идея свопа бд, но она не решает проблему качественно, только улучшит количественный предел спама в бд, который рдо выдержит.

**Comment by aurusov** [comment]
> В таком случае неправильное понимание работы IDE не у меня, а у разработчиков eclipse - потому что я могу руками написать такой код на джаве и eclipse зависнет.

Всё она умеет

``` java
package qqq;

public class Qqq {
    public static void main(String[] args) {
        while (true) {
        }
    }
}
```

`Console/Terminate` (красная кнопка в окне консоли)

> Своей `IDE`-то у нас нет теперь.

Даже если бы Эклипс не умел, то пользователю мы предлагаем IDE. Все остальное - наши проблемы.

Идеи
- Мы точно умеем останавливать бесконечный цикл. УМЕЕМ, Карл. Не теряй эту мысль.
- Надо понять почему в случае БД такое происходит. Если это из-за `GC`, то надо научиться
  - Перехватывать его сообщения о проблемах и стопить модель с ошибкой
  - Научиться правильно готовить `GC` для хранения большого кол-ва маленьких объектов

**Comment by bogachev-pa** [comment]
Бесконечный цикл тут ни при чем. Умирает `gc` из-за слишком большого количества объектов в памяти. Точно такое же поведение вы можете наблюдать, если запустите обычную модель на очень большое время, включив сериализацию. Если заглянете во время этого "зависания" в родительский эклипс, то сможете увидеть (возможно, через какое-то время) сообщения в `stderr` в духе `OutOfMemoryError: GC overhead limit exceeded`.

Теперь касательно того, как с этим бороться. Я правильно понимаю, что под `GB` вы имели в виду сборщик мусора?

**Comment by aurusov** [comment]
Да, опечатался.

**Comment by aurusov** [comment]
Предлагается добавить watch-dog для GC.
